File: framework/fit/java/fit-builtin/plugins/fit-service-coordination-nacos/src/main/java/modelengine/fit/service/server/NacosRegistryServer.java
/*
 * Copyright (c) 2025 Huawei Technologies Co., Ltd. All rights reserved.
 * This file is a part of the ModelEngine Project.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */

package modelengine.fit.service.server;

import static com.alibaba.nacos.api.naming.PreservedMetadataKeys.HEART_BEAT_INTERVAL;
import static com.alibaba.nacos.api.naming.PreservedMetadataKeys.HEART_BEAT_TIMEOUT;
import static modelengine.fitframework.inspection.Validation.notBlank;
import static modelengine.fitframework.inspection.Validation.notNull;

import com.alibaba.nacos.api.exception.NacosException;
import com.alibaba.nacos.api.naming.NamingFactory;
import com.alibaba.nacos.api.naming.NamingService;
import com.alibaba.nacos.api.naming.listener.EventListener;
import com.alibaba.nacos.api.naming.listener.NamingEvent;
import com.alibaba.nacos.api.naming.pojo.Instance;
import com.alibaba.nacos.api.naming.pojo.ListView;
import com.alibaba.nacos.client.naming.listener.NamingChangeEvent;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import modelengine.fit.service.Notify;
import modelengine.fit.service.RegistryService;
import modelengine.fit.service.entity.Address;
import modelengine.fit.service.entity.Application;
import modelengine.fit.service.entity.ApplicationInstance;
import modelengine.fit.service.entity.Endpoint;
import modelengine.fit.service.entity.FitableAddressInstance;
import modelengine.fit.service.entity.FitableInfo;
import modelengine.fit.service.entity.FitableMeta;
import modelengine.fit.service.entity.FitableMetaInstance;
import modelengine.fit.service.entity.GenericableInfo;
import modelengine.fit.service.entity.Worker;
import modelengine.fitframework.annotation.Component;
import modelengine.fitframework.annotation.Fitable;
import modelengine.fitframework.conf.Config;
import modelengine.fitframework.conf.runtime.CommunicationProtocol;
import modelengine.fitframework.conf.runtime.MatataConfig;
import modelengine.fitframework.conf.runtime.WorkerConfig;
import modelengine.fitframework.log.Logger;
import modelengine.fitframework.util.ObjectUtils;
import modelengine.fitframework.util.StringUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Properties;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Service for providing Nacos registry center functionality.
 *
 * @author 董智豪
 * @since 2025-06-04
 */
@Component
public class NacosRegistryServer implements RegistryService {
    private static final Logger log = Logger.get(NacosRegistryServer.class);
    private static final String CLUSTER_DOMAIN_KEY = "cluster.domain";
    private static final Pattern CLUSTER_PORT_PATTERN = Pattern.compile("cluster\\.(.*?)\\.port");
    private static final String WORKER_KEY = "worker";
    private static final String APPLICATION_KEY = "application";
    private static final String FITABLE_META_KEY = "fitable-meta";
    private static final String SEPARATOR = "::";

    private final NamingService namingService;
    private final ObjectMapper objectMapper = new ObjectMapper();
    private final NacosConfig nacosConfig;
    private final MatataConfig matata;
    private final Notify notify;
    private final WorkerConfig worker;
    private final Map<String, com.alibaba.nacos.api.naming.listener.EventListener> serviceSubscriptions =
            new ConcurrentHashMap<>();

    public NacosRegistryServer(Notify notify, WorkerConfig worker, Config config,MatataConfig matata)
            throws NacosException {
        notNull(config, "The configuration cannot be null.");
        this.matata = notNull(matata, "The matata configuration cannot be null.");
        this.notify = notNull(notify, "The registry listener cannot be null.");
        this.worker = notNull(worker, "The worker config cannot be null.");
        this.nacosConfig = config.get("matata.registry.nacos", NacosConfig.class);
        this.namingService = NamingFactory.createNamingService(getNacosProperties());
    }

    private Properties getNacosProperties() {
        Properties properties = new Properties();
        String serverAddr = this.matata.registry().host() + ":" + this.matata.registry().port();
        notBlank(serverAddr, "The Nacos server address cannot be blank.");
        properties.put("serverAddr", serverAddr);
        properties.put("username", ObjectUtils.nullIf(this.nacosConfig.getUsername(), StringUtils.EMPTY));
        properties.put("password", ObjectUtils.nullIf(this.nacosConfig.getPassword(), StringUtils.EMPTY));
        properties.put("namespace", ObjectUtils.nullIf(this.matata.registry().environment(), StringUtils.EMPTY));
        properties.put("accessKey", ObjectUtils.nullIf(this.nacosConfig.getAccessKey(), StringUtils.EMPTY));
        properties.put("secretKey", ObjectUtils.nullIf(this.nacosConfig.getSecretKey(), StringUtils.EMPTY));
        return properties;
    }

    /**
     * Builds a unique key in the format {@code <groupName>::<serviceName>} for {@code <serviceSubscriptions>}.
     *
     * @param groupName The group name as {@link String}.
     * @param serviceName The service name as {@link String}.
     * @return A concatenated key like {@code groupName::serviceName}.
     */
    private String buildServiceKey(String groupName, String serviceName) {
        return groupName + SEPARATOR + serviceName;
    }

    @Override
    @Fitable(id = "register-fitables")
    public void registerFitables(List<FitableMeta> fitableMetas, Worker worker, Application application) {
        try {
            log.debug("Registering fitables. [fitableMetas={}, worker={}, aplication={}]",
                    fitableMetas,
                    worker.getId(),
                    application.getNameVersion());
            for (FitableMeta meta : fitableMetas) {
                FitableInfo fitable = meta.getFitable();
                String groupName = this.getGroupName(fitable);
                String serviceName = this.getServiceName(fitable);
                List<Instance> instances = createInstance(worker, application, meta);
                for (Instance instance : instances) {
                    this.namingService.registerInstance(serviceName, groupName, instance);
                }
            }
        } catch (NacosException e) {
            log.error("Failed to register fitables due to registry error.", e);
        }
    }

    private String getServiceName(FitableInfo fitable) {
        return fitable.getFitableId() + SEPARATOR + fitable.getFitableVersion();
    }

    private String getGroupName(FitableInfo fitable) {
        return fitable.getGenericableId() + SEPARATOR + fitable.getGenericableVersion();
    }

    private String getGroupName(GenericableInfo genericable) {
        return genericable.getGenericableId() + SEPARATOR + genericable.getGenericableVersion();
    }

    private List<Instance> createInstance(Worker worker, Application application, FitableMeta meta) {
        log.debug("Creating instance for worker. [worker={}, application={}, meta={}]",
                worker.getId(),
                application.getNameVersion(),
                meta);
        List<Instance> instances = new ArrayList<>();
        for (Address address : worker.getAddresses()) {
            List<Endpoint> endpoints = address.getEndpoints();
            for (Endpoint endpoint : endpoints) {
                Instance instance = new Instance();
                instance.setIp(address.getHost());
                instance.setPort(endpoint.getPort());
                HashMap<String, String> metadata = this.buildInstanceMetadata(worker, application, meta);
                instance.setMetadata(metadata);
                this.setInstanceProperties(instance);
                instances.add(instance);
            }
        }
        return instances;
    }

    /**
     * Build metadata for service instance, including worker, application and FitableMeta information.
     *
     * @param worker The worker node object.
     * @param application The application object.
     * @param meta The {@link FitableMeta} metadata object.
     * @return A {@link Map} containing all serialized metadata.
     */
    private HashMap<String, String> buildInstanceMetadata(Worker worker, Application application, FitableMeta meta) {
        HashMap<String, String> metadata = new HashMap<>();
        if (this.nacosConfig.getHeartbeatInterval() != null) {
            metadata.put(HEART_BEAT_INTERVAL, String.valueOf(this.nacosConfig.getHeartbeatInterval()));
        }
        if (this.nacosConfig.getHeartbeatTimeout() != null) {
            metadata.put(HEART_BEAT_TIMEOUT, String.valueOf(this.nacosConfig.getHeartbeatTimeout()));
        }
        try {
            metadata.put(WORKER_KEY, this.objectMapper.writeValueAsString(worker));
            metadata.put(APPLICATION_KEY, this.objectMapper.writeValueAsString(application));
            metadata.put(FITABLE_META_KEY, this.objectMapper.writeValueAsString(meta));
        } catch (JsonProcessingException e) {
            log.error("Failed to serialize metadata for worker.", e);
        }
        return metadata;
    }

    private void setInstanceProperties(Instance instance) {
        if (!this.nacosConfig.getIsEphemeral()) {
            instance.setEphemeral(false);
        }
        if (this.nacosConfig.getWeight() != null) {
            instance.setWeight(this.nacosConfig.getWeight());
        }
    }

    @Override
    @Fitable(id = "unregister-fitables")
    public void unregisterFitables(List<FitableInfo> fitables, String workerId) {
        log.debug("Unregistering fitables for worker. [fitables={}, workerId={}]", fitables, workerId);
        for (FitableInfo fitable : fitables) {
            this.unregisterSingleFitable(fitable, workerId);
        }
    }

    private void unregisterSingleFitable(FitableInfo fitable, String workerId) {
        String groupName = this.getGroupName(fitable);
        String serviceName = this.getServiceName(fitable);
        try {
            List<Instance> instances = this.namingService.selectInstances(serviceName, groupName, true);
            this.unregisterMatchingInstances(instances, workerId, serviceName, groupName);
        } catch (NacosException e) {
            log.error("Failed to unregister fitable due to registry error.", e);
        }
    }

    private void unregisterMatchingInstances(List<Instance> instances, String workerId, String serviceName,
            String groupName) {
        for (Instance instance : instances) {
            try {
                Worker worker = this.objectMapper.readValue(instance.getMetadata().get(WORKER_KEY), Worker.class);
                if (Objects.equals(workerId, worker.getId())) {
                    this.namingService.deregisterInstance(serviceName, groupName, instance);
                }
            } catch (JsonProcessingException e) {
                log.error("Failed to parse worker metadata for fitable.", e);
            } catch (NacosException e) {
                log.error("Failed to deregister instance.", e);
            }
        }
    }

    @Override
    @Fitable(id = "query-fitables-addresses")
    public List<FitableAddressInstance> queryFitables(List<FitableInfo> fitables, String workerId) {
        log.debug("Querying fitables for worker. [fitables={}, workerId={}]", fitables, workerId);
        Map<FitableInfo, FitableAddressInstance> resultMap = new HashMap<>();
        for (FitableInfo fitable : fitables) {
            try {
                List<Instance> instances = this.queryInstances(fitable);
                if (instances.isEmpty()) {
                    continue;
                }
                this.processApplicationInstances(resultMap, fitable, instances);
            } catch (Exception e) {
                log.error("Failed to query fitables for genericableId.", e);
            }
        }
        return new ArrayList<>(resultMap.values());
    }

    private void processApplicationInstances(Map<FitableInfo, FitableAddressInstance> resultMap, FitableInfo fitable,
            List<Instance> instances) {
        Map<Application, List<Instance>> appInstancesMap = groupInstancesByApplication(instances);
        for (Map.Entry<Application, List<Instance>> entry : appInstancesMap.entrySet()) {
            Application application = entry.getKey();
            List<Instance> appInstances = entry.getValue();
            FitableMeta meta = parseFitableMeta(appInstances.get(0));
            Set<Worker> workers = extractWorkers(appInstances, application);
            FitableAddressInstance fai = resultMap.computeIfAbsent(fitable, k -> {
                FitableAddressInstance newFai = new FitableAddressInstance();
                newFai.setFitable(fitable);
                newFai.setApplicationInstances(new ArrayList<>());
                return newFai;
            });
            ApplicationInstance appInstance = new ApplicationInstance();
            appInstance.setApplication(application);
            appInstance.setFormats(meta.getFormats());
            appInstance.setWorkers(new ArrayList<>(workers));
            fai.getApplicationInstances().add(appInstance);
        }
    }

    /**
     * Extract all workers corresponding to instances and adjust addresses based on application extension information.
     *
     * @param appInstances The list of application instances.
     * @param application The application object.
     * @return Set of workers.
     */
    private Set<Worker> extractWorkers(List<Instance> appInstances, Application application) {
        Set<Worker> workers = new HashSet<>();
        for (Instance instance : appInstances) {
            Worker worker = parseWorker(instance);
            workers.add(worker);
        }
        if (application.getExtensions().containsKey(CLUSTER_DOMAIN_KEY)) {
            this.replaceAddresses(workers, application);
        }
        return workers;
    }

    private Map<Application, List<Instance>> groupInstancesByApplication(List<Instance> instances) {
        Map<Application, List<Instance>> map = new HashMap<>();
        for (Instance instance : instances) {
            Application app = this.parseApplication(instance);
            map.computeIfAbsent(app, k -> new ArrayList<>()).add(instance);
        }
        return map;
    }

    private List<Instance> queryInstances(FitableInfo fitable) throws NacosException {
        String groupName = this.getGroupName(fitable);
        String serviceName = this.getServiceName(fitable);
        return this.namingService.selectInstances(serviceName, groupName, true);
    }

    private FitableMeta parseFitableMeta(Instance instance) {
        try {
            return this.objectMapper.readValue(instance.getMetadata().get(FITABLE_META_KEY), FitableMeta.class);
        } catch (JsonProcessingException e) {
            log.error("Failed to parse fitable meta for instance.", e);
            FitableMeta meta = new FitableMeta();
            meta.setFitable(new FitableInfo());
            return meta;
        }
    }

    private Application parseApplication(Instance instance) {
        try {
            return this.objectMapper.readValue(instance.getMetadata().get(APPLICATION_KEY), Application.class);
        } catch (JsonProcessingException e) {
            log.error("Failed to parse application metadata for instance.", e);
            Application app = new Application();
            app.setNameVersion("unknown");
            return app;
        }
    }

    private Worker parseWorker(Instance instance) {
        try {
            return this.objectMapper.readValue(instance.getMetadata().get(WORKER_KEY), Worker.class);
        } catch (JsonProcessingException e) {
            log.error("Failed to parse worker metadata for instance.", e);
            Worker worker = new Worker();
            Address address = new Address();
            address.setHost(instance.getIp());

            Endpoint endpoint = new Endpoint();
            endpoint.setPort(instance.getPort());
            endpoint.setProtocol(1);

            address.setEndpoints(Collections.singletonList(endpoint));
            worker.setAddresses(Collections.singletonList(address));
            return worker;
        }
    }

    private void replaceAddresses(Set<Worker> workers, Application application) {
        Address address = new Address();
        address.setHost(application.getExtensions().get(CLUSTER_DOMAIN_KEY));
        address.setEndpoints(buildEndPoints(application.getExtensions()));
        workers.forEach(w -> w.setAddresses(Collections.singletonList(address)));
    }

    private List<Endpoint> buildEndPoints(Map<String, String> extensions) {
        List<Endpoint> endpoints = new ArrayList<>();
        for (Map.Entry<String, String> entry : extensions.entrySet()) {
            Matcher matcher = CLUSTER_PORT_PATTERN.matcher(entry.getKey());
            if (matcher.matches()) {
                String protocolName = matcher.group(1);
                CommunicationProtocol protocol = CommunicationProtocol.valueOf(StringUtils.toUpperCase(protocolName));
                Endpoint endpoint = new Endpoint();
                endpoint.setPort(Integer.valueOf(entry.getValue()));
                endpoint.setProtocol(protocol.code());
                endpoints.add(endpoint);
            }
        }
        return endpoints;
    }

    @Override
    @Fitable(id = "subscribe-fitables")
    public List<FitableAddressInstance> subscribeFitables(List<FitableInfo> fitables, String workerId,
            String callbackFitableId) {
        log.debug("Subscribing to fitables for worker. [fitables={}, workerId={}, callbackFitableId={}]",
                fitables,
                workerId,
                callbackFitableId);
        for (FitableInfo fitable : fitables) {
            try {
                String groupName = this.getGroupName(fitable);
                String serviceName = this.getServiceName(fitable);
                if (this.serviceSubscriptions.containsKey(buildServiceKey(groupName, serviceName))) {
                    log.debug("Already subscribed to service. [groupName={}, serviceName={}]", groupName, serviceName);
                    continue;
                }
                EventListener eventListener =
                        this.serviceSubscriptions.computeIfAbsent(buildServiceKey(groupName, serviceName),
                                k -> event -> {
                                    if (event instanceof NamingEvent || event instanceof NamingChangeEvent) {
                                        onServiceChanged(fitable);
                                    }
                                });
                this.namingService.subscribe(serviceName, groupName, eventListener);
            } catch (NacosException e) {
                log.error("Failed to subscribe to Nacos service.", e);
            }
        }
        return this.queryFitables(fitables, workerId);
    }

    @Override
    @Fitable(id = "unsubscribe-fitables")
    public void unsubscribeFitables(List<FitableInfo> fitables, String workerId, String callbackFitableId) {
        log.debug("Unsubscribing from fitables for worker. [fitables={}, workerId={}, callbackFitableId={}]",
                fitables,
                workerId,
                callbackFitableId);
        for (FitableInfo fitable : fitables) {
            try {
                String groupName = this.getGroupName(fitable);
                String serviceName = this.getServiceName(fitable);
                EventListener listener = this.serviceSubscriptions.get(buildServiceKey(groupName, serviceName));
                this.namingService.unsubscribe(serviceName, groupName, listener);
                this.serviceSubscriptions.remove(buildServiceKey(groupName, serviceName));
            } catch (NacosException e) {
                log.error("Failed to unsubscribe from Nacos service.", e);
            }
        }
    }

    /**
     * Handle service change events, query and notify updates to Fitables instance information.
     *
     * @param fitableInfo The changed Fitables information.
     */
    private void onServiceChanged(FitableInfo fitableInfo) {
        List<FitableAddressInstance> fitableAddressInstances =
                this.queryFitables(Collections.singletonList(fitableInfo), this.worker.id());
        this.notify.notifyFitables(fitableAddressInstances);
    }

    @Override
    @Fitable(id = "query-running-fitables")
    public List<FitableMetaInstance> queryFitableMetas(List<GenericableInfo> genericables) {
        log.debug("Querying fitable metas for genericables. [genericables={}]", genericables);
        Map<FitableMeta, Set<String>> metaEnvironments = new HashMap<>();

        for (GenericableInfo genericable : genericables) {
            this.processGenericableServices(genericable, metaEnvironments);
        }

        return this.buildFitableMetaInstances(metaEnvironments);
    }

    private void processGenericableServices(GenericableInfo genericable,
            Map<FitableMeta, Set<String>> metaEnvironments) {
        String groupName = this.getGroupName(genericable);
        try {
            ListView<String> services = this.namingService.getServicesOfServer(1, Integer.MAX_VALUE, groupName);
            for (String serviceName : services.getData()) {
                this.processServiceInstances(serviceName, groupName, metaEnvironments);
            }
        } catch (NacosException e) {
            log.error("Failed to query fitable metas.", e);
        }
    }

    private void processServiceInstances(String serviceName, String groupName,
            Map<FitableMeta, Set<String>> metaEnvironments) {
        try {
            List<Instance> instances = this.namingService.selectInstances(serviceName, groupName, true);
            if (instances.isEmpty()) {
                return;
            }
            FitableMeta meta = parseFitableMeta(instances.get(0));
            this.collectEnvironmentsFromInstances(instances, meta, metaEnvironments);
        } catch (NacosException e) {
            log.error("Failed to select instances for service: " + serviceName, e);
        }
    }

    private void collectEnvironmentsFromInstances(List<Instance> instances, FitableMeta meta,
            Map<FitableMeta, Set<String>> metaEnvironments) {
        for (Instance instance : instances) {
            try {
                Worker worker = this.objectMapper.readValue(instance.getMetadata().get(WORKER_KEY), Worker.class);
                metaEnvironments.computeIfAbsent(meta, k -> new HashSet<>()).add(worker.getEnvironment());
            } catch (JsonProcessingException e) {
                log.error("Failed to parse worker metadata.", e);
            }
        }
    }

    private List<FitableMetaInstance> buildFitableMetaInstances(Map<FitableMeta, Set<String>> metaEnvironments) {
        List<FitableMetaInstance> results = new ArrayList<>();
        for (Map.Entry<FitableMeta, Set<String>> entry : metaEnvironments.entrySet()) {
            FitableMetaInstance instance = new FitableMetaInstance();
            instance.setMeta(entry.getKey());
            instance.setEnvironments(new ArrayList<>(entry.getValue()));
            results.add(instance);
        }
        return results;
    }
}

File: framework/fit/python/fit_py_nacos_registry/service/nacos_registry_server.py
# -- encoding: utf-8 --
# Copyright (c) 2025 Huawei Technologies Co., Ltd. All Rights Reserved.
# This file is a part of the ModelEngine Project.
# Licensed under the MIT License. See License.txt in the project root for license information.
# ======================================================================================================================
"""
Service for providing Nacos registry center functionality.
"""
from concurrent.futures import ThreadPoolExecutor
from typing import List, Dict
import weakref
import atexit

from v2.nacos import RegisterInstanceParam, ListInstanceParam, \
    DeregisterInstanceParam, SubscribeServiceParam, Instance, ListServiceParam

from fitframework import fitable, const
from fitframework.api.logging import sys_plugin_logger
from fit_common_struct.entity import Worker, FitableMeta, Application, FitableAddressInstance, \
    FitableMetaInstance, ApplicationInstance
from fit_common_struct.core import Fitable, Genericable

from .config import get_nacos_namespace
from .utils import build_service_key, get_service_name, get_group_name_from_fitable, \
    get_group_name_from_genericable, create_instances
from .parsers import parse_fitable_meta, parse_application, parse_worker, \
    group_instances_by_application, extract_workers
from .async_executor import run_async_safely, call_list_instances, call_deregister_instance, \
    call_subscribe, call_unsubscribe, call_list_services, call_register_instance

# Global variables
_service_subscriptions: weakref.WeakValueDictionary = weakref.WeakValueDictionary()
_executor = ThreadPoolExecutor(max_workers=10)


def _cleanup_executor():
    """Cleanup the thread pool executor."""
    try:
        if _executor:
            _executor.shutdown(wait=True)
            sys_plugin_logger.info("Thread pool executor shut down successfully")
    except Exception as e:
        sys_plugin_logger.error(f"Error shutting down thread pool executor: {e}")


# Register cleanup function to ensure executor is properly closed
atexit.register(_cleanup_executor)


def on_service_changed(fitable_info: Fitable, worker_id: str) -> None:
    """
    Handle service change events, query and notify updates to Fitables instance information.

    Args:
        fitable_info: The changed Fitables information.
        worker_id: The worker ID.
    """
    try:
        # Query current instances
        instances = query_fitable_addresses([fitable_info], worker_id)
        # notify_fitables(instances)
        sys_plugin_logger.debug(
            f"Service changed for fitable: {fitable_info}, instances: {len(instances)}"
        )
    except Exception as e:
        sys_plugin_logger.error(f"Service change handling failed: {e}")


def extract_workers(app_instances: List[Instance], application: Application) -> List[Worker]:
    """
    Extract all workers corresponding to instances.

    Args:
        app_instances: The list of application instances.
        application: The application object.

    Returns:
        List of workers.
    """
    workers = []
    for instance in app_instances:
        worker = parse_worker(instance)
        workers.append(worker)

    return workers


@fitable(const.REGISTER_FIT_SERVICE_GEN_ID, const.REGISTER_FIT_SERVICE_FIT_ID)
def register_fitables(fitable_metas: List[FitableMeta], worker: Worker, application: Application) -> None:
    """
    Register Fitable service implementations to the registry center.

    Args:
        fitable_metas: List of Fitable metadata to register.
        worker: Current FIT process information.
        application: Current application information.

    Raises:
        Exception: If registration fails due to registry error.
    """
    try:
        sys_plugin_logger.debug(
            f"Registering fitables. [fitableMetas={fitable_metas}, "
            f"worker={worker.id}, application={application.nameVersion}]"
        )

        for meta in fitable_metas:
            fitable = meta.fitable
            group_name = get_group_name_from_fitable(fitable)
            service_name = get_service_name(fitable)

            instances = create_instances(worker, application, meta)
            for instance in instances:
                param = RegisterInstanceParam(
                    service_name=service_name,
                    group_name=group_name,
                    ip=instance["ip"],
                    port=instance["port"],
                    weight=instance["weight"],
                    ephemeral=instance["ephemeral"],
                    metadata=instance["metadata"]
                )
                run_async_safely(call_register_instance(param))

        sys_plugin_logger.info(f"Successfully registered fitables for worker {worker.id}")
    except Exception as e:
        sys_plugin_logger.error(f"Failed to register fitables due to registry error: {e}")
        raise

@fitable(const.UNREGISTER_FIT_SERVICE_GEN_ID, const.UNREGISTER_FIT_SERVICE_FIT_ID)
def unregister_fitables(fitables: List[Fitable], worker_id: str) -> None:
    """
    Unregister service implementations from the registry center.

    Args:
        fitables: List of Fitable implementations to unregister.
        worker_id: Unique identifier of the process where service implementations reside.
    """
    sys_plugin_logger.debug(
        f"Unregistering fitables for worker. [fitables={fitables}, workerId={worker_id}]"
    )

    for fitable in fitables:
        unregister_single_fitable(fitable, worker_id)


def unregister_single_fitable(fitable: Fitable, worker_id: str) -> None:
    """
    Unregister a single Fitable implementation.

    Args:
        fitable: The Fitable implementation to unregister.
        worker_id: The worker ID.
    """
    group_name = get_group_name_from_fitable(fitable)
    service_name = get_service_name(fitable)

    try:
        # Get all instances for the service
        param = ListInstanceParam(
            service_name=service_name,
            group_name=group_name,
            healthy_only=True
        )
        instances = run_async_safely(call_list_instances(param))
        unregister_matching_instances(instances, worker_id, service_name, group_name)
    except Exception as e:
        sys_plugin_logger.error(f"Failed to unregister fitable due to registry error: {e}")


def unregister_matching_instances(instances: List[Instance], worker_id: str, service_name: str, group_name: str) -> None:
    """
    Unregister matching instances.

    Args:
        instances: List of instances to check.
        worker_id: The worker ID to match.
        service_name: The service name.
        group_name: The group name.
    """
    for instance in instances:
        try:
            worker = parse_worker(instance)
            if worker and worker.id == worker_id:
                param = DeregisterInstanceParam(
                    service_name=service_name,
                    group_name=group_name,
                    ip=instance.ip,
                    port=instance.port
                )
                run_async_safely(call_deregister_instance(param))
                sys_plugin_logger.debug(f"Successfully deregistered instance {instance.ip}:{instance.port}")
        except Exception as e:
            sys_plugin_logger.error(f"Failed to deregister instance: {e}")


@fitable(const.QUERY_FIT_SERVICE_GEN_ID, const.QUERY_FIT_SERVICE_FIT_ID)
def query_fitable_addresses(fitables: List[Fitable], worker_id: str) -> List[FitableAddressInstance]:
    """
    Query instance information for Fitable implementations (pull mode).

    Args:
        fitables: List of Fitable implementation information.
        worker_id: Current FIT process identifier.

    Returns:
        List of obtained instance information.
    """
    sys_plugin_logger.debug(
        f"Querying fitables for worker. [fitables={fitables}, workerId={worker_id}]"
    )
    result_map = {}

    for fitable in fitables:
        try:
            instances = query_instances(fitable)
            if not instances:
                continue
            process_application_instances(result_map, fitable, instances)
        except Exception as e:
            sys_plugin_logger.error(f"Failed to query fitables for genericableId: {e}")

    return list(result_map.values())


def query_instances(fitable: Fitable) -> List[Instance]:
    """
    Query instances for a specific Fitable.

    Args:
        fitable: The Fitable to query instances for.

    Returns:
        List of instances.
    """
    group_name = get_group_name_from_fitable(fitable)
    service_name = get_service_name(fitable)

    param = ListInstanceParam(
        service_name=service_name,
        group_name=group_name,
        healthy_only=True
    )
    return run_async_safely(call_list_instances(param))


def process_application_instances(result_map: Dict, fitable: Fitable, instances: List[Instance]) -> None:
    """
    Process application instances and group them.

    Args:
        result_map: Dictionary to store results.
        fitable: The Fitable being processed.
        instances: List of instances to process.
    """
    app_instances_map = group_instances_by_application(instances)
    
    for app, app_instances in app_instances_map.items():
        meta = parse_fitable_meta(app_instances[0].metadata)
        workers = extract_workers(app_instances, app)
        
        fai = result_map.get(fitable)
        if fai is None:
            fai = FitableAddressInstance(applicationInstances=[], fitable=fitable)
            result_map[fitable] = fai
        
        app_instance = ApplicationInstance(workers=list(workers), application=app, formats=meta.formats if meta.formats else [])
        fai.applicationInstances.append(app_instance)


@fitable(const.SUBSCRIBE_FIT_SERVICE_GEN_ID, const.SUBSCRIBE_FIT_SERVICE_FIT_ID)
def subscribe_fit_service(fitables: List[Fitable], worker_id: str, callback_fitable_id: str) -> List[FitableAddressInstance]:
    """
    Subscribe to Fitable service instance information (push mode).

    Args:
        fitables: List of Fitable implementation information.
        worker_id: Current FIT process identifier.
        callback_fitable_id: Identifier for callback Fitable implementation.

    Returns:
        Queried instance information.
    """
    sys_plugin_logger.debug(
        f"Subscribing to fitables for worker. [fitables={fitables}, "
        f"workerId={worker_id}, callbackFitableId={callback_fitable_id}]"
    )

    # Register subscriptions
    for fitable in fitables:
        try:
            group_name = get_group_name_from_fitable(fitable)
            service_name = get_service_name(fitable)
            service_key = build_service_key(group_name, service_name)

            if service_key in _service_subscriptions:
                sys_plugin_logger.debug(
                    f"Already subscribed to service. [groupName={group_name}, serviceName={service_name}]"
                )
                continue

            # Create event listener
            def create_event_listener(fitable_ref: Fitable, worker_id_ref: str):
                def event_listener(event):
                    _executor.submit(on_service_changed, fitable_ref, worker_id_ref)
                return event_listener

            event_listener = create_event_listener(fitable, worker_id)
            _service_subscriptions[service_key] = event_listener

            # Register subscription
            param = SubscribeServiceParam(
                service_name=service_name,
                group_name=group_name,
                subscribe_callback=event_listener
            )
            run_async_safely(call_subscribe(param))
            sys_plugin_logger.debug(
                f"Subscribed to service. [groupName={group_name}, serviceName={service_name}]"
            )

        except Exception as e:
            sys_plugin_logger.error(f"Failed to subscribe to Nacos service: {e}")

    return query_fitable_addresses(fitables, worker_id)


@fitable(const.UNSUBSCRIBE_FIT_SERVICE_GEN_ID, const.UNSUBSCRIBE_FIT_SERVICE_FIT_ID)
def unsubscribe_fitables(fitables: List[Fitable], worker_id: str, callback_fitable_id: str) -> None:
    """
    Unsubscribe from specified Fitable service instance information.

    Args:
        fitables: List of specified Fitable implementations.
        worker_id: Unique identifier of the specified process.
        callback_fitable_id: Unique identifier for unsubscribe callback Fitable implementation.
    """
    sys_plugin_logger.debug(f"Unsubscribing from fitables for worker. [fitables={fitables}, workerId={worker_id}, callbackFitableId={callback_fitable_id}]")
    
    for fitable in fitables:
        try:
            group_name = get_group_name_from_fitable(fitable)
            service_name = get_service_name(fitable)
            service_key = build_service_key(group_name, service_name)

            # Use pop with default to avoid KeyError if listener was garbage collected
            listener = _service_subscriptions.pop(service_key, None)
            if listener is not None:
                param = SubscribeServiceParam(
                    service_name=service_name,
                    group_name=group_name,
                    subscribe_callback=listener
                )
                run_async_safely(call_unsubscribe(param))
                sys_plugin_logger.debug(f"Unsubscribed from service. [groupName={group_name}, serviceName={service_name}]")
            else:
                sys_plugin_logger.debug(f"Listener already cleaned up for service. [groupName={group_name}, serviceName={service_name}]")
        except Exception as e:
            sys_plugin_logger.error(f"Failed to unsubscribe from Nacos service: {e}")


@fitable(const.QUERY_FITABLE_METAS_GEN_ID, const.QUERY_FITABLE_METAS_FIT_ID)
def query_fitable_metas(genericable_infos: List[Genericable]) -> List[FitableMetaInstance]:
    """
    Query Fitable metadata from the registry center.

    Args:
        genericable_infos: List of Genericable information.

    Returns:
        List of queried Fitable metadata.
    """
    sys_plugin_logger.debug(
        f"Querying fitable metas for genericables. [genericables={genericable_infos}]"
    )
    meta_environments = {}

    for genericable in genericable_infos:
        process_genericable_services(genericable, meta_environments)

    return build_fitable_meta_instances(meta_environments)


def process_genericable_services(genericable: Genericable, meta_environments: Dict) -> None:
    """
    Process service list for a Genericable.

    Args:
        genericable: The Genericable to process.
        meta_environments: Dictionary to collect metadata environments.
    """
    group_name = get_group_name_from_genericable(genericable)

    try:
        # Get all services under the group
        param = ListServiceParam(
            namespace_id=get_nacos_namespace(),
            group_name=group_name,
            page_no=1,
            page_size=1000  # Assume fetching enough services at once
        )
        service_list = run_async_safely(call_list_services(param))

        for service_name in service_list.services:
            process_service_instances(service_name, group_name, meta_environments)
    except Exception as e:
        sys_plugin_logger.error(f"Failed to query fitable metas: {e}")


def process_service_instances(service_name: str, group_name: str, meta_environments: Dict) -> None:
    """
    Process service instances.

    Args:
        service_name: The service name.
        group_name: The group name.
        meta_environments: Dictionary to collect metadata environments.
    """
    try:
        # Get service instances
        param = ListInstanceParam(
            service_name=service_name,
            group_name=group_name,
            healthy_only=True
        )
        instances = run_async_safely(call_list_instances(param))

        if not instances:
            return

        meta = parse_fitable_meta(instances[0].metadata)
        collect_environments_from_instances(instances, meta, meta_environments)
    except Exception as e:
        sys_plugin_logger.error(f"Failed to select instances for service {service_name}: {e}")


def collect_environments_from_instances(instances: List[Instance], meta: FitableMeta, meta_environments: Dict) -> None:
    """
    Collect environment information from instances.

    Args:
        instances: List of instances to process.
        meta: FitableMeta object.
        meta_environments: Dictionary to collect environments.
    """
    for instance in instances:
        try:
            worker = parse_worker(instance)
            if worker and worker.environment:
                if meta not in meta_environments:
                    meta_environments[meta] = set()
                meta_environments[meta].add(worker.environment)
        except Exception as e:
            sys_plugin_logger.error(f"Failed to parse worker metadata: {e}")


def build_fitable_meta_instances(meta_environments: Dict) -> List[FitableMetaInstance]:
    """
    Build FitableMetaInstance list.

    Args:
        meta_environments: Dictionary mapping metadata to environments.

    Returns:
        List of FitableMetaInstance objects.
    """
    results = []
    for meta, envs in meta_environments.items():
        instance = FitableMetaInstance(meta, list(envs))
        results.append(instance)
    return results
